src/sec_tpm.erl

```diff
@@ -0,0 +1,182 @@
+-define(TPM2_QUOTE_CMD,
+	"tpm2_quote -Q --key-context " ++ ?AK_CTX_FILE ++ " -l sha256:0,1 --message " ++ ?QUOTE_MSG_FILE ++ " --signature " ++
+		?QUOTE_SIG_FILE ++ " --qualification "
+).
+-define(TPM2_VERIFY_CMD,
+	"tpm2_verifysignature -c " ++ ?AK_CTX_FILE ++ " -g sha256 -m " ++ ?QUOTE_MSG_FILE ++ " -s " ++ ?QUOTE_SIG_FILE
+).
+
+%% Generate an attestation using the provided nonce (as a string)
+generate_attestation(Nonce) ->
+	% Check if the root directory exists, and create it if not
+	case filelib:is_dir(?ROOT_DIR) of
+		true -> ok;
+		false -> file:make_dir(?ROOT_DIR)
+	end,
+
+	% Setup the keys
+	setup_keys(),
+
+	% Use the address in hex format as the nonce
+	NonceHex = binary_to_list(binary:encode_hex(Nonce)),
+	Command = ?TPM2_QUOTE_CMD ++ NonceHex,
+	?c({"Running command", Command}),
+	case sec_helpers:run_command(Command) of
+		{ok, _} ->
+			% Read the quote.msg, quote.sig, and ak.pub files
+			{_, QuoteData} = sec_helpers:read_file(?QUOTE_MSG_FILE),
+			{_, SignatureData} = sec_helpers:read_file(?QUOTE_SIG_FILE),
+			{_, AkPubData} = sec_helpers:read_file(?AK_PUB_FILE),
+
+			% Ensure the read data is in binary format
+			QuoteBin = list_to_binary(QuoteData),
+			SignatureBin = list_to_binary(SignatureData),
+			AkPubBin = list_to_binary(AkPubData),
+
+			% Get sizes of the individual files (in binary)
+			QuoteSize = byte_size(QuoteBin),
+			SignatureSize = byte_size(SignatureBin),
+			AkPubSize = byte_size(AkPubBin),
+
+			% Create a binary header with the sizes and offsets
+			Header = <<QuoteSize:32/unit:8, SignatureSize:32/unit:8, AkPubSize:32/unit:8>>,
+
+			% Create a binary with all three files' data concatenated after the header
+			AttestationBinary = <<Header/binary, QuoteBin/binary, SignatureBin/binary, AkPubBin/binary>>,
+
+			% Log the data (if you need to debug)
+			?c({"Attestation generated, binary data:", AttestationBinary}),
+
+			% Return the binary containing the header and files
+			{ok, AttestationBinary};
+		{error, Reason} ->
+			?c({"Error generating attestation", Reason}),
+			{error, Reason}
+	end.
+
+%% Verify the attestation using the AttestationBinary
+verify_attestation(AttestationBinary) ->
+	% Extract the header (size info)
+	<<QuoteSize:32/unit:8, SignatureSize:32/unit:8, AkPubSize:32/unit:8, Rest/binary>> = AttestationBinary,
+
+	% Extract the individual components using the sizes from the header
+	<<QuoteData:QuoteSize/binary, Rest1/binary>> = Rest,
+	<<SignatureData:SignatureSize/binary, Rest2/binary>> = Rest1,
+	<<AkPubData:AkPubSize/binary>> = Rest2,
+
+	% Write the components to temporary files (if needed for verification)
+	sec_helpers:write_to_file(?QUOTE_MSG_FILE, QuoteData),
+	sec_helpers:write_to_file(?QUOTE_SIG_FILE, SignatureData),
+	sec_helpers:write_to_file(?AK_PUB_FILE, AkPubData),
+
+	% Run the TPM verification command using the files
+	CommandVerify = ?TPM2_VERIFY_CMD,
+	?c({"Running command", CommandVerify}),
+	case sec_helpers:run_command(CommandVerify) of
```

Comment:
As I see it run command returns {ok, Output} in all cases. 
```
run_command(Command) ->
	?c({"Executing command", Command}),
	Output = os:cmd(Command ++ " 2>&1"),
	case Output of
		% Empty output interpreted as success if no output is expected
		"" -> {ok, []};
		% Return output for further inspection
		_ -> {ok, Output}
	end.

```

Could we have a test case that shows how it fails?

Reply:
Discussed with @oltarasenko marking as resolved.